### PR TITLE
Reduce server freezes during large explosions

### PR DESCRIPTION
**Pull request by Lohikar**
Moved explosion calculations into a process that processes a queue of explosions with `SCHECK`s in areas with significant processing.
Significantly reduces server freezes during large explosions, and almost completely nullifies them for small ones.

Likely subtly changes how explosions behave. 

Powernet rebuilds should be looked at too.

### CODE REVIEW

File: code/controllers/Processes/explosives.dm [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+		ticks_without_work++
+		if (ticks_without_work > 5)
+			if(defer_powernet_rebuild)
+				makepowernets()
+				defer_powernet_rebuild = 0
+		return
+
+	ticks_without_work = 0
+
+	for (var/datum/explosiondata/data in work_queue)
```
Comment (skull132) [inline code/controllers/Processes/explosives.dm]: Minor optimization point.

Since you're expecting work_queue to only contain datum/explosion data, you can forego a istype() check and typecast implicitly:
```
for (var/A in work_queue)
    var/datum/explosiondata/data = A
```

File: code/controllers/Processes/explosives.dm [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+				Array.sense_explosion(x0,y0,z0,devastation_range,heavy_impact_range,light_impact_range,took)
+
+		work_queue -= data
+
+	if(!powernet_rebuild_was_deferred_already && defer_powernet_rebuild)
+		makepowernets()
+		defer_powernet_rebuild = 0
+
+
+/datum/controller/process/explosives/proc/queue(var/data)
```
Comment (skull132) [inline code/controllers/Processes/explosives.dm]: Maybe just in case make this run istype() and !data checks? That way you ensure in the integrity of work_queue, without bogging down the main loop.

**Comment by skull132** [comment]
A few more tricks that other servers use for explosions are to freeze the lighting and atmospherics processors while an explosion is in progress. Though it produced mixed results when I tested it some 7 months ago.

It might be an idea to freeze them only for larger explosions.

File: code/controllers/Processes/explosives.dm [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+		var/devastation_range = data.devastation_range
+		var/heavy_impact_range = data.heavy_impact_range
+		var/light_impact_range = data.light_impact_range
+		var/flash_range = data.flash_range
+		var/adminlog = data.adminlog
+		var/z_transfer = data.z_transfer
+
+		var/power = max(0,devastation_range) * 2 + max(0,heavy_impact_range) + max(0,light_impact_range)
+		if(config.use_recursive_explosions)
+			explosion_rec(epicenter, power)
```
Comment (skull132) [inline code/controllers/Processes/explosives.dm]: This is still unprotected from lag. Should also do that, because some procs go straight to this.

**Comment by NanakoAC** [comment]
Please preserve my additions of localised explosion audio and screen shaking. see how they're interwoven with the explosion code

**Comment by skull132** [comment]
They are preserved AFAIK, he copy-pastad the explosion proc over.

Anyways. I did some research. For **science** and all. And I discovered an inherent flaw in the way we use processors and in the way SCHECK is written. The issue is here:
```
#define SCHECK_EVERY(this_many_calls) if(++src.calls_since_last_scheck >= this_many_calls) sleepCheck()
#define SCHECK SCHECK_EVERY(50)
```

Basically, it means that SCHECK only even _considers_ sleeping every 50 calls to it for the same processor. Which will basically permit any processor to run over tick (and thus cause lag) somewhere in those 50 calls. My suspicion is that it's a left over from the old, initial implementation of Goon's processScheduler, which used a more time consuming .dll call and time comparison in sleepCheck, when compared to the really basic world.tick_usage checks we use now.

On my test server, I altered #SCHECK to look like this: `#define SCHECK SCHECK_EVERY(0)` and was able to explode the SM without any lag, with no further modification necessary. I've also implemented this alteration on the live server to monitor how it'll affect the processors on varying degrees of realistic load. My fear is that it may cause some processors to lag behind too much, though that doesn't seem to be happening.

In summary, along with making recursive booms adhere to the controller (new proc perhaps) I'd like you to remove SCHECK_EVERY define, and change SCHECK to simply be a macro for sleepCheck(). This will enable all controllers to sleep precisely when they need to, and will pretty much kill lag from processor controlled stacks.

**Comment by NanakoAC** [comment]
reducing freezes is good, but shouldn't it be possible to completely and utterly eliminate freezing from even the largest of explosions, if done correctly?

**Comment by skull132** [comment]
@NanakoAC read my blurb. That basically does it, minus edge cases that would have to be hunted down.

File: code/controllers/Processes/explosives.dm [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+	var/turf/epicenter = data.epicenter
+	var/devastation_range = data.devastation_range
+	var/heavy_impact_range = data.heavy_impact_range
+	var/light_impact_range = data.light_impact_range
+	var/flash_range = data.flash_range
+	var/adminlog = data.adminlog
+	var/z_transfer = data.z_transfer
+
+	var/power = max(0,devastation_range) * 2 + max(0,heavy_impact_range) + max(0,light_impact_range)
+	if(config.use_recursive_explosions)
```
Comment (skull132) [inline code/controllers/Processes/explosives.dm]: This can now be removed.

File: code/controllers/Processes/explosives.dm [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+/datum/controller/process/explosives/proc/explosion(var/datum/explosiondata/data)
+	var/turf/epicenter = data.epicenter
+	var/devastation_range = data.devastation_range
+	var/heavy_impact_range = data.heavy_impact_range
+	var/light_impact_range = data.light_impact_range
+	var/flash_range = data.flash_range
+	var/adminlog = data.adminlog
+	var/z_transfer = data.z_transfer
+
+	var/power = max(0,devastation_range) * 2 + max(0,heavy_impact_range) + max(0,light_impact_range)
```
Comment (skull132) [inline code/controllers/Processes/explosives.dm]: Duplication of operations. This data bit is already saved as data.rec_pow.

File: code/__defines/machinery.dm [hunk 6]
```diff
@@ -1,4 +1,4 @@
-var/global/defer_powernet_rebuild = 0      // True if net rebuild will be called manually after an event.
+//var/global/defer_powernet_rebuild = 0      // True if net rebuild will be called manually after an event.
```
Comment (skull132) [inline code/__defines/machinery.dm]: Please just delete it. There's no point in keeping commented out vars about.
